cpp/src/copying/contiguous_split.cu

```diff
@@ -1158,116 +1277,735 @@ std::vector<packed_table> contiguous_split(cudf::table_view const& input,
+        size_of_batches_per_iteration.push_back(iter_size_bytes);
+        accum_size_per_iteration.push_back(accum_size);
+
+        if (next_iteration_it == h_offsets.end()) { break; }
+
+        current_offset_it += batches_in_iter;
+      }
+    }
+
+    // apply changed offset
+    {
+      rmm::device_uvector<std::size_t> d_accum_size_per_iteration(
+        accum_size_per_iteration.size(), stream, mr);
+
+      CUDF_CUDA_TRY(cudaMemcpyAsync(d_accum_size_per_iteration.data(),
+                                    accum_size_per_iteration.data(),
+                                    accum_size_per_iteration.size() * sizeof(std::size_t),
+                                    cudaMemcpyDefault,
+                                    stream.value()));
+
+      // we want to update the offset of batches for every iteration, except the first one (because
+      // offsets in the first iteration are all 0 based)
+      auto num_batches_in_first_iteration = num_batches_per_iteration[0];
+      auto iter           = thrust::make_counting_iterator(num_batches_in_first_iteration);
+      auto num_iterations = accum_size_per_iteration.size();
+      thrust::for_each(
+        rmm::exec_policy(stream, mr),
+        iter,
+        iter + num_batches - num_batches_in_first_iteration,
+        [num_iterations,
+         d_batched_dst_buf_info     = d_batched_dst_buf_info.begin(),
+         d_accum_size_per_iteration = d_accum_size_per_iteration.begin()] __device__(size_type i) {
+          auto prior_iteration_size =
+            thrust::upper_bound(thrust::seq,
+                                d_accum_size_per_iteration,
+                                d_accum_size_per_iteration + num_iterations,
+                                d_batched_dst_buf_info[i].dst_offset) -
+            1;
+          d_batched_dst_buf_info[i].dst_offset -= *prior_iteration_size;
+        });
+    }
+    return std::make_unique<chunk_iteration_state>(std::move(d_batched_dst_buf_info),
+                                                   std::move(d_batch_offsets),
+                                                   std::move(num_batches_per_iteration),
+                                                   std::move(size_of_batches_per_iteration),
+                                                   accum_size);
+
+  } else {
+    // we instantiate an "iteration state" for the regular single pass contiguous_split
+    // consisting of 1 iteration with all of the batches and totalling `total_size` bytes.
+    auto const total_size = std::reduce(h_buf_sizes, h_buf_sizes + num_partitions);
+
+    // 1 iteration with the whole size
+    return std::make_unique<chunk_iteration_state>(
+      std::move(d_batched_dst_buf_info),
+      std::move(d_batch_offsets),
+      std::move(std::vector<std::size_t>{static_cast<std::size_t>(num_batches)}),
+      std::move(std::vector<std::size_t>{total_size}),
+      total_size);
+  }
+}
+
+// template<bool T>
```

Comment:
will remove this, I was trying other approaches to passing a functor to `copy_partitions` to find the destination buffer pointer.